### PR TITLE
Major CI refactor

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
- building steps are now put into script
- do not publish on PRs
- namespace is all janitortechnology/
- persists ubuntu-dev to workspace

### CODE REVIEW

File: .circleci/build_image.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
+set -e
+
+if [[ $# -ne 2 ]]; then
+  echo "Usage: $0 DOCKERFILE IMAGE_NAME"
+  echo ".dockerfile is appended to DOCKERFILE."
+  echo "janitortechnology/ is prepended to IMAGE_NAME."
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: Please don't append or prepend anything to parameters. It makes things opaque and confusing (e.g. "What is chromium/chromium?") and it reduces flexibility (e.g. what if we want to build an importer Dockerfile that doesn't have a `.dockerfile` extension? Or what if we want to tag some image with a different prefix? Etc.)

**Comment by ishitatsuyuki** [comment]
One thing I noticed is that we may be facing endless PR builds filling the queue. Probably we should add --cache-from for other images as well?

File: .circleci/build_image.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  fi
+else
+  DOCKERHUB_TAG="pr"
+fi
+
+if [[ $CIRCLE_JOB = "ubuntu-dev" ]]; then
+  # Pull published `ubuntu-dev` image for cached layers. (Allowed to fail)
+  docker pull janitortechnology/ubuntu-dev:latest || true
+else
+  # Load the `ubuntu-dev` base image from previous step.
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: I believe that the correct Circle CI nomenclature is "job" instead of "step".

File: .circleci/build_image.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+if [[ $CIRCLE_JOB = "ubuntu-dev" ]]; then
+  docker build --cache-from=janitortechnology/ubuntu-dev:latest -t janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -f "$DOCKERFILE.dockerfile" .
+else
+  docker build -t janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -f "$DOCKERFILE.dockerfile" .
+fi
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  docker push $IMAGE_NAME:$DOCKERHUB_TAG
+fi
+
+if [[ $CIRCLE_JOB = "ubuntu-dev" ]]; then
```
Comment (ishitatsuyuki) [inline .circleci/build_image.sh]: Forgot to cd back.

**Comment by ishitatsuyuki** [comment]
I disagree that we should put the scripts inside the yaml. The size of the whole script is pretty large to be copy pasted for each jobs, and unfortunately yaml anchors cannot be used as well because it's a list not mapping.
Editing such long script inside yaml is a pain: you get excessive indents, and miss syntax highlighting for bash. This also makes modifying the script harder; instead of using variables, they're hardcoded and insanely hard to be replaced.

Also, having a separate deploy step sounds good at first. However, the drawback is that you now need to persist much more images, taking the risk of hitting size limit, and as well incurs network overhead.

To sum up, I think putting everything inside a script is not a bad thing. We're not hiding anything, it's just a piece of readable and maintainable code.

**Comment by jankeromnes** [comment]
Thank you for explaining your concerns. We both want what's best for the project, and I'm confident we can reach a shared understanding of the situation that will lead us to the best solution to enable CI for pull requests.

> The size of the whole script is pretty large to be copy pasted for each jobs

It is indeed very large right now, but I feel that's mostly due to the fact that we attempt to do many things conditionally, while these things could be avoided (the complicated `DOCKERHUB_TAG`) or rolled into each job where they make sense (e.g. instead of writing `if [[ $CIRCLE_JOB = "ubuntu-dev ]]` inside each job, we just add the relevant instruction to the `ubuntu-dev` job, and not in any other job).

> Editing such long script inside yaml is a pain

Again, I don't think the in-yaml bash commands would be as long as the original script. If you look at my suggested job setup and workflow, it's always 3 or 4 command lines at most per job, and they're extremely easy to read, understand and modify if needed.

> Also, having a separate deploy step sounds good at first. However, the drawback is that you now need to persist much more images, taking the risk of hitting size limit, and as well incurs network overhead.

That's a worrying concern. I don't know what the cache size limit is, or if there are network usage limits on Circle CI.

However, I still think it's worth trying the simplest solution first, and then react to any problems/limitations if they arise, instead of falling into the trap of premature optimization.

**Comment by ishitatsuyuki** [comment]
> Again, I don't think the in-yaml bash commands would be as long as the original script. If you look at my suggested job setup and workflow, it's always 3 or 4 command lines at most per job, and they're extremely easy to read, understand and modify if needed.

3 or 4 is already long enough. There are lots of repeated words where you can easily miss one. They're easy to read, but not easy to modify. You get 8x amplification for modifying the script, and 4x amplification for changing/adding a new name.

The same applies to a separate deploy step. You need to copy two more blocks for each new image. From a contributor's perspective, this cannot be really called simple.

(By the way, as an alternative, CircleCI parallelism, or Travis matrix allows you to achieve job generation with more minimum yaml. The parallelism has fixed count, so it may be suboptimal for this pattern though.)

File: .circleci/build_image.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  fi
+else
+  DOCKERHUB_TAG="pr"
+fi
+
+# Pull published image for cached layers. (Allowed to fail)
+docker pull janitortechnology/$IMAGE_NAME:latest || true
+if [[ $CIRCLE_JOB != "ubuntu-dev" ]]; then
+  # Load the `ubuntu-dev` base image from previous step.
+  docker load -i ubuntu-dev.tar
```
Comment (ishitatsuyuki) [inline .circleci/build_image.sh]: We should execute this earlier so we don't need to pull ubuntu-dev from the Hub.

**Comment by jankeromnes** [comment]
> 3 or 4 is already long enough. There are lots of repeated words where you can easily miss one. They're easy to read, but not easy to modify.

I agree that we shouldn't make mistakes in our Circle CI job commands. That's why we have code review, and it's pretty obvious when problems still slip through, because Circle CI will tell us about it (and it won't break anything in production, so it's not too dramatic).

I argue that 8 blocks of similar 3 or 4 lines are fairly easy to refactor (even though we don't expect to do that very often), e.g. with vim macros, or multi-cursor, or similar multi-editing tool.

> You get 8x amplification for modifying the script, and 4x amplification for changing/adding a new name.

I agree that a script allows to factor out similarity between all the jobs. While we should mind the trade-off between aggressively de-duplicating all the commands, and keeping all the code easy to read and to understand by anyone, I'm not fundamentally opposed to having a script.

> The same applies to a separate deploy step. You need to copy two more blocks for each new image. From a contributor's perspective, this cannot be really called simple.

Actually, people aren't supposed to contribute to this repository directly. Ideally, to add a new project to Janitor, a contributor would write a development Dockerfile based on `ubuntu-dev`, and host it in the project's own repository (that's what Kresus and PeerTube already do, and eventually I'd like all or most of the dockerfiles in this repository to be migrated away into their respective project's repository).

So the number of projects in your `circle.yml` is supposed to be constant, and shrink over time.

File: .circleci/build_image.sh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+docker build --cache-from=janitortechnology/$IMAGE_NAME:latest -t janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -f "$DOCKERFILE.dockerfile" .
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  docker push $IMAGE_NAME:$DOCKERHUB_TAG
+fi
+cd ..
+
+if [[ $CIRCLE_JOB = "ubuntu-dev" ]]; then
+  # Save the `ubuntu-dev` base image for next step.
+  docker save janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -o ubuntu-dev.tar
+fi
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: Please remove this `if [[ $CIRCLE_JOB = "ubuntu-dev" ]]` block from the common script, and add the `docker save` instruction directly to the `ubuntu-dev` job in `circle.yml` (no point in having it in a script that's common to all jobs).

File: .circleci/build_image.sh [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  # Log in to Docker Hub.
+  docker login -u $DOCKER_USER -p $DOCKER_PASS
+  # Decide whether Docker Hub images should be tagged ":branch-Y" or ":latest".
+  if [[ $CIRCLE_BRANCH != "master" ]]; then
+    DOCKERHUB_TAG="branch-$CIRCLE_BRANCH"
+  else
+    DOCKERHUB_TAG="latest"
+  fi
+else
+  DOCKERHUB_TAG="pr"
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: Please completely remove the `DOCKERHUB_TAG` variable. I think we should either publish images as `:latest` (i.e. the default value, so we can omit the tag) when we're on `master` branch, or not publish them at all.

Comment (ishitatsuyuki) [reply .circleci/build_image.sh]: I don't agree with that. There are cases where we want to test new images, such as images utilizing Docker volumes.

Comment (ishitatsuyuki) [reply .circleci/build_image.sh]: I have briefly updated things, but the image save/load depends on `DOCKERHUB_TAG` so it's blocked on this discussion regarding `DOCKERHUB_TAG`.

**Comment by ishitatsuyuki** [comment]
I think I have addressed the concerns. Please give another review.

**Comment by ishitatsuyuki** [comment]
Re-triggering a build.

**Comment by ishitatsuyuki** [comment]
Hmm, the trick didn't work. @jankeromnes can you push the rebuild button, and review afterwards?

**Comment by jankeromnes** [comment]
> @jankeromnes can you push the rebuild button

Sure! Done. Thanks a lot for all your hard work on this! I'll try to find some time to review this soon (I'm very sorry about my review lag).

File: .circleci/build_image.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  # Decide whether Docker Hub images should be tagged ":branch-Y" or ":latest".
+  if [[ $CIRCLE_BRANCH != "master" ]]; then
+    DOCKERHUB_TAG="branch-$CIRCLE_BRANCH"
+  else
+    DOCKERHUB_TAG="latest"
+  fi
+fi
+
+# Pull published image for cached layers. (Allowed to fail)
+docker pull $IMAGE_NAME || true
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: I don't think we should `docker pull` here:

- In the previous Circle CI configuration, we only pulled `janx/ubuntu-dev` to pass the image from one job to the next, and you're replacing this with `docker save`/`load`.
- We purposefully *didn't* cache project images, in order to ensure that builds continue to work from scratch and that repositories/sources are always the freshest possible.
- (We also pulled `ubuntu-dev` specifically in the `ubuntu-dev` job, in order to accelerate base image builds, but I'm ok with removing that call as well, in order to get fresher packages as well.)

File: .circleci/build_image.sh [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+DIR=`dirname $1`
+DOCKERFILE=$1
+IMAGE_NAME=$2
+
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  # Log in to Docker Hub.
+  docker login -u $DOCKER_USER -p $DOCKER_PASS
+  # Decide whether Docker Hub images should be tagged ":branch-Y" or ":latest".
+  if [[ $CIRCLE_BRANCH != "master" ]]; then
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: This `DOCKERHUB_TAG` variable is only used to tag the image after it was built. Please move this `if [[ $CIRCLE_BRANCH != "master" ]]` block into the `if [[ -z $CIRCLE_PULL_REQUEST ]]` at the bottom of this file.

File: .circleci/build_image.sh [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  exit 2
+fi
+
+DIR=`dirname $1`
+DOCKERFILE=$1
+IMAGE_NAME=$2
+
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  # Log in to Docker Hub.
+  docker login -u $DOCKER_USER -p $DOCKER_PASS
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Also, I wonder if this `docker login` call couldn't also be moved down, just before the `docker tag`&`push` (assuming that we remove the `docker pull` call). That way, we'd only have a single `if [[ -z $CIRCLE_PULL_REQUEST ]]` block instead of two.

**Comment by ishitatsuyuki** [comment]
Addressed the comments. I've swapped out the debloat commits (as it's still pending review), but keep in mind that there was some cases where a too big ubuntu-dev can fail to be saved.

File: .circleci/build_image.sh [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+if [[ $# -ne 2 ]]; then
+  echo "Usage: $0 DOCKERFILE IMAGE_NAME"
+  exit 2
+fi
+
+DIR=`dirname $1`
+DOCKERFILE=$1
+IMAGE_NAME=$2
+
+docker build --cache-from=$IMAGE_NAME -t $IMAGE_NAME -f $DOCKERFILE $DIR
```
Comment (jankeromnes) [inline .circleci/build_image.sh]: Nit: I believe this should be `--cache-from=janitortechnology/ubuntu-dev`. (But do we even need to specify `--cache-from` after we `docker load -i ubuntu-dev.tar`?)

Comment (ishitatsuyuki) [reply .circleci/build_image.sh]: Removed. Thanks!
